### PR TITLE
Docerina support for record/object field-level documentation

### PR DESCRIPTION
**Pull request by pramodya1994**
Ballerina:
```ballerina
# Person record
public type Person record {|
    # name of the person
    string name;
    # age of the person
    int age;
    # country code of the person
    CountryCode countryCode;
|};
```

API Docs:
Screenshot 2020-03-30 at 15 31 19

If field description is given in the record/object Documentation and field-level both, priority is given to the record/object Documentation.
```ballerina
# Apartment record
#
# + number - number of the apartment 
# + street - street of the apartment 
# + countryCode - country code of the apartment 
public type Appartment record {|
    # apartment no
    int number;
    # apartment street
    string street;
    # apartment country-code
    CountryCode countryCode;
|};
```

API Docs:
Screenshot 2020-03-30 at 15 33 41

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/22307

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/docerina/FieldLevelDocsTest.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        Map<String, ModuleDoc> docsMap = BallerinaDocGenerator.generateModuleDocsFromBLangPackages(
+                Paths.get("src/test/resources", sourceRoot).toAbsolutePath().toString(), modules);
+        List<ModuleDoc> moduleDocList = new ArrayList<>(docsMap.values());
+        moduleDocList.sort(Comparator.comparing(pkg -> pkg.bLangPackage.packageID.toString()));
+        Map<String, List<Path>> resources = new HashMap<>();
+
+        Project project = BallerinaDocGenerator.getDocsGenModel(moduleDocList, resources);
+        testModule = project.modules.get(0);
+
+        List<Record> records = testModule.records;
```
Comment (keizer619) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/docerina/FieldLevelDocsTest.java]: @pramodya1994 Do we need to declare these variable if its only used once?

File: misc/docerina/src/main/java/org/ballerinalang/docgen/docs/BallerinaDocGenerator.java [hunk 2]
```diff
@@ -209,7 +179,15 @@ public static void writeAPIDocsForModules(Map<String, ModuleDoc> docsMap, String
 
         String rootPathModuleLevel = project.isSingleFile ? "./" : "../";
         String rootPathConstructLevel = project.isSingleFile ? "../" : "../../";
+
         // Generate module pages
+        if (project.modules == null) {
+            String errMessage =
+                    "docerina: API documentation generation failed. Couldn't create the [output directory] " + output;
+            out.println(errMessage);
```
Comment (keizer619) [inline misc/docerina/src/main/java/org/ballerinalang/docgen/docs/BallerinaDocGenerator.java]: @pramodya1994 We need to remove this println

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/docerina/FieldLevelDocsTest.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+    @BeforeClass
+    public void setup() throws IOException {
+        String sourceRoot = "test-src/docerina/fieldLevelDocsProject";
+        CompileResult result = BCompileUtil.compile(sourceRoot, "testModule");
+
+        List<BLangPackage> modules = new LinkedList<>();
+        modules.add((BLangPackage) result.getAST());
+        Map<String, ModuleDoc> docsMap = BallerinaDocGenerator.generateModuleDocsFromBLangPackages(
+                Paths.get("src/test/resources", sourceRoot).toAbsolutePath().toString(), modules);
```
Comment (keizer619) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/docerina/FieldLevelDocsTest.java]: @pramodya1994 is above path works on windows?

File: tests/jballerina-unit-test/src/test/resources/test-src/docerina/fieldLevelDocsProject/src/testModule/test.bal [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+# + name - student name
+# + age - student age
+public type Student object {
+    public string name = "Saman";
+    public int age = 15;
+};
+
+# Teacher object
+public type Teacher object {
+    # Teacher name
```
Comment (hevayo) [inline tests/jballerina-unit-test/src/test/resources/test-src/docerina/fieldLevelDocsProject/src/testModule/test.bal]: Shall we have some test cases with md at field level

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/docerina/FieldLevelDocsTest.java (hunk 1 above)
Comment (pramodya1994) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/docerina/FieldLevelDocsTest.java]: Removed

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/docerina/FieldLevelDocsTest.java (hunk 3 above)
Comment (pramodya1994) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/docerina/FieldLevelDocsTest.java]: Corrected

File: tests/jballerina-unit-test/src/test/resources/test-src/docerina/fieldLevelDocsProject/src/testModule/test.bal (hunk 4 above)
Comment (pramodya1994) [reply tests/jballerina-unit-test/src/test/resources/test-src/docerina/fieldLevelDocsProject/src/testModule/test.bal]: Added a test case

**Comment by pubudu91** [comment]
Why was priority given to the field description at object/record def level? Shouldn't the priority be given for the field level doc? That's the one specified in the spec and the current way of documenting fields in the object/record doc is a convention we came up with.

**Comment by pramodya1994** [comment]
> Why was priority given to the field description at object/record def level? Shouldn't the priority be given for the field level doc? That's the one specified in the spec and the current way of documenting fields in the object/record doc is a convention we came up with.

Added a separate issue to track and resolve this
https://github.com/ballerina-platform/ballerina-lang/issues/22402

**Pull request merged**